### DISCUSSION TITLE
adding a base record for exception

### DISCUSSION THREAD

**Discussion by Mr0N** [dotnet/csharplang#6671] [Language Ideas]
I suggest adding a base record for exception
Similar to this https://github.com/Mr0N/CustomExceptionRecords

**Comment by ghost** [comment]
I couldn't figure out the best area label to add to this issue. If you have write-permissions please help me learn by adding exactly one [area label](https://github.com/dotnet/runtime/blob/master/docs/area-owners.md).

**Comment by antomys** [comment]
Well, it doesn't make sense. Frankly speaking, you reinvented `System.Exception` class.

**Comment by Mr0N** [comment]
> Well, it doesn't make sense. Frankly speaking, you reinvented `System.Exception` class.

for record

**Comment by antomys** [comment]
Could you define, what is a major difference between `record` and `class`?

**Comment by stephentoub** [comment]
It sounds like what you're really asking for is the ability to declare a record that derives from Exception.  That's an issue for dotnet/csharplang, the ability to derive a record from a non-record.

**Comment by Mr0N** [comment]
> Could you define, what is a major difference between `record` and `class`?

The record can be written much faster and can be wtites into one file.

**Comment by Mr0N** [comment]
> It sounds like what you're really asking for is the ability to declare a record that derives from Exception. That's an issue for dotnet/csharplang, the ability to derive a record from a non-record.

This is a runtime problem, because I implemented it without changing the programming language.

**Comment by stephentoub** [comment]
> This is a runtime problem, because I implemented it without changing the programming language.

RecordException would need to derive from Exception. And if that were possible, you wouldn't need RecordException at all, as any record could derive from Exception directly.

"record" is a construct of the C# language. Augmenting its design and capabilities are the domain of the C# language.

**Comment by antomys** [comment]
> It sounds like what you're really asking for is the ability to declare a record that derives from an Exception. That's an issue for `dotnet/csharplang`, the ability to derive a record from a non-record.

Oh, I see. Thanks for elaborating. So the main suggestion of this issue shortens to "Add ability to inherit a record from a non-record (here in issue: record from class)).  So, as I see it, this issue is not an issue, but a suggestion. Not sure that it's going to be useful, but anyway

**Comment by Mr0N** [comment]
> > This is a runtime problem, because I implemented it without changing the programming language.
> 
> RecordException would need to derive from Exception. And if that were possible, you wouldn't need RecordException at all, as any record could derive from Exception directly.
> 
> "record" is a construct of the C# language. Augmenting its design and capabilities are the domain of the C# language.



> > This is a runtime problem, because I implemented it without changing the programming language.
> 
> RecordException would need to derive from Exception. And if that were possible, you wouldn't need RecordException at all, as any record could derive from Exception directly.
> 
> "record" is a construct of the C# language. Augmenting its design and capabilities are the domain of the C# language.

Well,  change the logic of work "throw" and make a basic type for excluding records

**Comment by Mr0N** [comment]
![image](https://user-images.githubusercontent.com/61237140/200594658-c297e1a9-061a-4a2c-bcfc-4e26e9d86655.png)

**Comment by Mr0N** [comment]
> > It sounds like what you're really asking for is the ability to declare a record that derives from an Exception. That's an issue for `dotnet/csharplang`, the ability to derive a record from a non-record.
> 
> Oh, I see. Thanks for elaborating. So the main suggestion of this issue shortens to "Add ability to inherit a record from a non-record (here in issue: record from class)). So, as I see it, this issue is not an issue, but a suggestion. Not sure that it's going to be useful, but anyway

Yes, I do it at the level of il code, but in my example it will not be possible to do this in C# code.

**Comment by stephentoub** [comment]
> Well, change the logic of work "throw" and make a basic type for excluding records

Again, that's the C# language. The runtime supports throwing anything.

This is not actionable here. I'm going to close it. Thanks.

**Comment by Mr0N** [comment]
> Reference

So my library changes the programming language :)

**Comment by HaloFour** [comment]
The limitation is that C# does not allow records to extend from non-records (other than `System.Object`), but all exceptions must extend from `System.Exception`.  I would imagine that if the language team were to revisit this they would do so through lifting the limitation on inheritance in general vs. doing something general purpose for exceptions.

See: https://github.com/dotnet/csharplang/blob/f856ab390d974896079905ef38479e00a54c5025/meetings/2020/LDM-2020-06-22.md#inheritance-between-records-and-classes

**Comment by Mr0N** [comment]
> 



> The limitation is that C# does not allow records to extend from non-records (other than `System.Object`), but all exceptions must extend from `System.Exception`. I would imagine that if the language team were to revisit this they would do so through lifting the limitation on inheritance in general vs. doing something general purpose for exceptions.
> 
> See: https://github.com/dotnet/csharplang/blob/f856ab390d974896079905ef38479e00a54c5025/meetings/2020/LDM-2020-06-22.md#inheritance-between-records-and-classes

The limit as it is now is normal, it is not easier to just write it in the il code as it is now, this is one class that will not change

**Comment by Mr0N** [comment]
And at the same time add the il code project

**Comment by HaloFour** [comment]
This is why I'm suggesting that if this is considered that it would be addressed by lifting the limitation, allowing records to extend from other arbitrary classes.  Then a record could extend from `System.Exception` directly.  It wouldn't be trivial, but it's much more likely to happen than to have a new base exception class.

**Comment by Mr0N** [comment]
> This is why I'm suggesting that if this is considered that it would be addressed by lifting the limitation, allowing records to extend from other arbitrary classes. Then a record could extend from `System.Exception` directly. It wouldn't be trivial, but it's much more likely to happen than to have a new base exception class.

It's bad, it's better to change the programming language, or to write one class in il, which is easier?

**Comment by HaloFour** [comment]
Either way the programming language would change because `RecordException` would have to derive from `Exception`, which is not legal in C#.  But by adding that one class all you've done is split the exception hierarchy since these record exceptions cannot extend from any of the common base exceptions defined in the BCL, like `ArgumentException` or `IOException`.

**Comment by Mr0N** [comment]
> 



> Either way the programming language would change because `RecordException` would have to derive from `Exception`, whic

Yes, but from this record, only records will be followed, it will have a base class Exception, the programming language at the C# level is not affected, because it is written in il code, a basic record for errors just appears.

**Comment by Mr0N** [comment]
> extend from any of the common base exceptions defined in the BCL, like `ArgumentException` or `IOException`.

I don't see the point of them imitating records, in fact a base class for records is needed to reduce the routine of writing classes for errors

**Comment by HaloFour** [comment]
> Yes, but from this record, only records will be followed, it will have a base class Exception, the programming language at the C# level is not affected, because it is written in il code, a basic record for errors just appears.

The BCL is written in C#, if C# doesn't allow it then it won't end up in the runtime.  They have already closed your suggestion of adding a new base exception class as something they won't do.

> I don't see the point of them imitating records, in fact a base class for records is needed to reduce the routine of writing classes for errors

Which just makes them that much less useful.

**Comment by Mr0N** [comment]
> The BCL is written in C#, if C# doesn't allow it then it won't end up in the runtime. They have already closed your suggestion of adding a new base exception class as something they won't do.
> 
> >

![image](https://user-images.githubusercontent.com/61237140/200610444-723ef6bc-58d2-494b-99b6-36af951e102c.png)

**Comment by HaloFour** [comment]
@Mr0N 

You'll also notice this:

<img width="441" alt="image" src="https://user-images.githubusercontent.com/10102290/200610781-a3fa1130-477c-46e5-9987-1100f4159658.png">

They've already stated that this is a language concern, and as a language concern the possible solution would be to lift the limitation on extending non-records.

The runtime team has expressed being not interested in writing custom IL for specific APIs that cannot be expressed in C#, such as defining helper methods that used `System.Delegate` or `System.Enum` constraints.

**Comment by Mr0N** [comment]
https://github.com/dotnet/runtime/blob/d459c65324aa3387252d480781e9c7ea6f934231/src/tests/JIT/opt/JitMinOpts/Perf/BBCnt1.il

**Comment by HaloFour** [comment]
That's a test.

This was transferred to the C# repo for a reason, it's a language issue.  If you're stuck on this being implemented as a special new class you're going to be disappointed.

**Comment by Mr0N** [comment]
> 



> @Mr0N
> 
> You'll also notice this:
> 
> <img alt="image" width="441" src="https://user-images.githubusercontent.com/10102290/200610781-a3fa1130-477c-46e5-9987-1100f4159658.png">
> 
> They've already stated that this is a language concern, and as a language concern the possible solution would be to lift the limitation on extending non-records.
> 
> The runtime team has expressed being not interested in writing custom IL for specific APIs that cannot be expressed in C#, such as defining helper methods that used `System.Delegate` or `System.Enum` constraints.

well, I don't know, well, maybe it's a problem of the programming language that needs to be implemented on the il code, just allow them to do it.

**Comment by stephentoub** [comment]
You're trying to use a feature of the C# language: records.  You don't like a limitation of that C# feature: records currently can't derive from non-records.  The fix isn't to special-case a single base type in the core libraries and hack it outside of C# to work-around the designed C# constraint for that one case.  You're asking for a custom Exception base type, but there are thousands and thousands of other classes that folks might similarly want as a base type for a record; Exception is not special in that regard.  Thank you for your suggestion, but we're not going to add such a type to the base libraries.

In the meantime, you've hacked up an approach in IL that obviously meets your needs, since you're asking us to replicate what you've done.  You're obviously free to use that as a helper in your own projects, and even to publish a NuGet package for others to consume if you believe it's so impactful.  In the meantime, I'd encourage you to have the conversation at the language level about lifting the constraint, and more importantly, how to address the holes that constraint is there to prevent.

Thanks.

**Comment by Mr0N** [comment]
> You're asking for a custom Exception base type, but there are thousands and thousands of other classes that folks might similarly want as a base type for a record;

well, they won't be able to do it (it's good, because these are fundamentally different things, I just don't see the second solution), this record will be an ordinary record, besides, it will be accepted by throw

**Comment by Mr0N** [comment]
Okay, this is just an idea, if you don’t want to implement it, I’ll use the library maybe)

**Reply by Mr0N** [reply]
Well, you can also add instead of "exception", like everything is the same as records, only it follows from the class of exceptions and is declared as the second word) only this language load is unnecessary)

**Reply by Mr0N** [reply]
myexception(string message):exception(message);

**Reply by HaloFour** [reply]
The language team has already expressed being willing to reconsider the imposed limitation given more use cases.  They have to work out the semantics of what it would mean, given records apply value identity which doesn't really work if the base class doesn't have value identity.  But that's not much worse than a record that has a field of a type that does not have value identity, IMO.

This use case with exception records is unlikely to case about value equality, which calls into question using records for these cases in general.  Other than a primary constructor, what exactly do you intend to be able to do with these exceptions?

**Reply by Mr0N** [reply]
> This use case with exception records is unlikely to case about value equality, which calls into question using records for these cases in general. Other than a primary constructor, what exactly do you intend to be able to do with these exceptions?

Well, the "exception" is just a class with  constructor in most cases, I just don't see the point in copying it every time

**Reply by HaloFour** [reply]
This sounds like a better use case for [primary constructors](https://github.com/dotnet/csharplang/issues/2691) then, not records.

**Reply by Mr0N** [reply]
> This sounds like a better use case for [primary constructors](https://github.com/dotnet/csharplang/issues/2691) then, not records.

I did not understand what it is, I doubt that it will be implemented, the records already exist.

**Reply by Mr0N** [reply]
![image](https://user-images.githubusercontent.com/61237140/200620157-1c497e67-e49d-4d67-bdd5-99005551e711.png)
do not like

**Reply by HaloFour** [reply]
Primary constructors are championed by the language team and it has been triaged into the working set.  It's a lot easier and more likely to happen then extending records to allow them to extend non-records, especially since you don't care about value equality.

**Reply by Mr0N** [reply]
> Primary constructors are championed by the language team and it has been triaged into the working set. It's a lot easier and more likely to happen then extending records to allow them to extend non-records, especially since you don't care about value equality.



Well, if it can be done that way, 
`record MyException(string message):RecordException(message);
`
and it will be done in one string,.and it will be in C# 12, then I'm in favor

**Reply by DavidArno** [reply]
> > Primary constructors are championed by the language team and it has been triaged into the working set. It's a lot easier and more likely to happen then extending records to allow them to extend non-records, especially since you don't care about value equality.
> 
> Well, if it can be done that way, `record MyException(string message):RecordException(message); ` and it will be done in one string,.and it will be in C# 12, then I'm in favor

No, it won't be done in that way as they won't be records. The syntax is still in discussion, but it will let you write something like:

```cs
class MyException(string message) : Exception(message);
```
This appears to be what you are asking for here: the ability to create a custom exception declaration on one line.

There's also no guarantee that it'll make it into C# 12. It might be in v13; it might never happen. There are no guarantees of what will make the cut for future versions of the language.
